sdk/data/azcosmos/cosmos_container_read_many.go

```diff
@@ -0,0 +1,130 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package azcosmos
+
+import (
+	"context"
+	"errors"
+	"runtime"
+	"sync"
+
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore"
+	"github.com/Azure/azure-sdk-for-go/sdk/data/azcosmos/queryengine"
+)
+
+// executeReadManyWithEngine executes a query using the provided query engine.
+func (c *ContainerClient) executeReadManyWithEngine(queryEngine queryengine.QueryEngine, items []ItemIdentity, readManyOptions *ReadManyOptions, operationContext pipelineRequestOptions, ctx context.Context) (ReadManyItemsResponse, error) {
+	// throw error that this is unsupported
+	return ReadManyItemsResponse{}, errors.New("ReadMany with query engine is not supported yet")
+}
+
+func (c *ContainerClient) executeReadManyWithPointReads(items []ItemIdentity, readManyOptions *ReadManyOptions, operationContext pipelineRequestOptions, ctx context.Context) (ReadManyItemsResponse, error) {
+
+	// if empty list of items, return empty list
+	if len(items) == 0 {
+		return ReadManyItemsResponse{}, nil
+	}
+	// Determine concurrency: use provided MaxConcurrency or number of CPU cores
+	var concurrency int
+	if readManyOptions != nil && readManyOptions.MaxConcurrency != nil && *readManyOptions.MaxConcurrency > 0 {
+		concurrency = int(*readManyOptions.MaxConcurrency)
+	} else {
+		concurrency = runtime.NumCPU()
+		if concurrency <= 0 {
+			concurrency = 1
+		}
+	}
+
+	// Prepare result slots to preserve input order
+	type slot struct {
+		value         []byte
+		requestCharge float32
+		err           error
+	}
+
+	results := make([]slot, len(items))
+
+	// Worker pool
+	var wg sync.WaitGroup
+	jobs := make(chan int)
+
+	// cancellation channel to short-circuit on first error
+	done := make(chan struct{})
+
+	// Start workers
+	workerCount := concurrency
+	if workerCount > len(items) {
+		workerCount = len(items)
+	}
+	itemOptions := ItemOptions{}
+	if readManyOptions != nil {
+		itemOptions.ConsistencyLevel = readManyOptions.ConsistencyLevel
+		itemOptions.SessionToken = readManyOptions.SessionToken
+	}
+	for worker := 0; worker < workerCount; worker++ {
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			for idx := range jobs {
+				select {
+				case <-done:
+					return
+				default:
+				}
+				item := items[idx]
+
+				itemResponse, err := c.ReadItem(ctx, item.PartitionKey, item.ID, &itemOptions)
+				if err != nil {
+					var azErr *azcore.ResponseError
+					// for 404, just continue without error
+					if errors.As(err, &azErr) {
+						if azErr.StatusCode == 404 {
+							return
```

Comment:
I think you want to continue the for loop rather than return. Returning would terminate this worker.